Fixes #2: Case Class Schema

thinkharderdev opened the pull request:
@jdegoes Can you take a look at this guy and see if this is what you had in mind?

Also, for implementing the Json decoder, can we assume that JSON keys are ordered consistently with the case class declaration. AFAICT this  assumption is made by zio-json but not sure if that is a correct assumption.

thinkharderdev commented:
JsonCodec is implemented. Still need to do the ProtobufCodec.

jdegoes commented on core/src/main/scala/zio/schema/Schema.scala:
```diff
@@ -27,6 +27,410 @@ object Schema {
+    field3: (String, Schema[A3]),
+    construct: (A1, A2, A3) => Z,
+    extract: Z => (A1, A2, A3)
+  ) extends CaseClass[Z] {
```
I think this encoding is almost perfect. After studying it for a while, I can see that `extract` will necessarily allocate a tuple. We can eliminate this overhead by switching to a slightly different encoding:

```scala
 case class CaseClass3[A1, A2, A3, Z](
     field1: (String, Schema[A1]),
     field2: (String, Schema[A2]),
     field3: (String, Schema[A3]),
     construct: (A1, A2, A3) => Z,
     extractField1: Z => A1,
     extractField2: Z => A2,
     extractField3: Z => A3
   ) extends CaseClass[Z] 
```

Then we could pick apart the `Z` into its component pieces without doing any allocations; and we can already, having all the component pieces, construct the `Z` by using `construct` without any allocations.

thinkharderdev replied:
I like this a lot. Should set us up well to implement Optics as well.

jdegoes commented on core/src/main/scala/zio/schema/Schema.scala:
```diff
@@ -27,6 +27,410 @@ object Schema {
+    override def toRecord: Record = Record(Map(field1, field2))
+  }
+
+  case class CaseClass3[A1, A2, A3, Z](
```
Here and elsewhere, I'd use:

```scala
final case class
```

Since it does not make sense to override case classes.

jdegoes commented on core/src/main/scala/zio/schema/Schema.scala:
```diff
@@ -27,6 +27,410 @@ object Schema {
+    construct: (A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16, A17, A18, A19, A20, A21, A22) => Z,
+    extract: Z => (A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16, A17, A18, A19, A20, A21, A22)
+  ) extends CaseClass[Z] {
+    override def toRecord: Record =
```
I love the `toRecord` on all of these, because it gives library authors a convenient way to generically handle all of them, albeit with performance penalty.

jdegoes commented on core/src/main/scala/zio/schema/Schema.scala:
```diff
@@ -27,6 +27,410 @@ object Schema {
   sealed case class Tuple[A, B](left: Schema[A], right: Schema[B])       extends Schema[(A, B)]
   final case class EitherSchema[A, B](left: Schema[A], right: Schema[B]) extends Schema[Either[A, B]]
 
+  sealed trait CaseClass[Z] extends Schema[Z] {
```
Making sure all the case class types extend this is perfect! This will allow more uniform, type safe handling in user-land code.

jdegoes commented on core/src/main/scala/zio/schema/codec/ProtobufCodec.scala:
```diff
@@ -383,6 +384,7 @@ object ProtobufCodec extends Codec {
         case Schema.Optional(codec)           => optionalDecoder(codec).asInstanceOf[Decoder[A]]
         case Schema.Fail(message)             => (_, _) => Left(message)
         case Schema.EitherSchema(left, right) => eitherDecoder(left, right).asInstanceOf[Decoder[A]]
+        case _                                => ???
```
Quick fix here and elsewhere will be `case caseClass : CaseClass[A] => decoder(caseClass.toRecord)`

thinkharderdev replied:
This won't work because you end up with a `Decoder[Map[String,_]]` instead of a `Decoder[A]`

I think it shouldn't be too hard to implement the protobuf encoder/decoder for the case class schemas though.

jdegoes commented:
@thinkharderdev This is great! A few minor comments but I am ready to merge when you're happy with it.

thinkharderdev commented:
@jdegoes Addressed review comments. Protobuf encoder/decoder is still unimplemented though. I can knock that out before we merge or I can do a follow-up PR if you'd rather get this in main

jdegoes commented:
@thinkharderdev I will merge now to avoid conflicts (it's a pretty big PR). Thanks for your work on this, and look forward to the follow-up PR!
